contracts/utils/Arrays.sol

```diff
@@ -12,6 +12,49 @@ import {Math} from "./math/Math.sol";
 library Arrays {
     using StorageSlot for bytes32;
 
+    /**
+     * @dev Sort an array (in memory) in increassing order.
+     *
+     * This function does the sorting "in place", meaning that it overrides the input. The object is returned for
+     * convenience, but that returned value can be discarded safely if the caller has a memory pointer to the array.
+     *
+     * Note: this function's cost is O(n.log(n)) in average and O(n²) in the worst case, with n the length of the
+     * array. Using it in view function that are executed through `eth_call` is safe, but one should be very careful
+     * when executing this as part of a transaction. Depending on the size of the array, this can be very exepensive
+     * to run, leading to potential DoS attacks.
+     */
+    function sort(uint256[] memory array) internal pure returns (uint256[] memory) {
+        _quickSort(array, 0, array.length);
+        return array;
+    }
+
+    function _quickSort(uint256[] memory array, uint256 i, uint256 j) private pure {
+        unchecked {
+            if (j - i < 2) return;
```

Comment:
This function is private. The user doesn't control the arguments `i` and `j`. The first call uses `0` and `array.length` and all recursive calls maintain the invariant that `j >= i`. So no risk of overflow here, and using uncheck saves gas in a function that is recursive (so executed a lot of times)

Reply:
[Skip to content](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#start-of-content)
https://github.com/OpenZeppelin
/
[openzeppelin-contracts](https://github.com/OpenZeppelin/openzeppelin-contracts)

Type / to search

[Code](https://github.com/OpenZeppelin/openzeppelin-contracts)
Issues
165
Pull requests
55
[Actions](https://github.com/OpenZeppelin/openzeppelin-contracts/actions)
Security
18
[Insights](https://github.com/OpenZeppelin/openzeppelin-contracts/pulse)
[Settings](https://github.com/OpenZeppelin/openzeppelin-contracts/settings)
Arrays sorting (in memory) #4846
 Open
[Amxx](https://github.com/Amxx) wants to merge 23 commits into [OpenZeppelin:master](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master) from [Amxx:feature/quicksort](https://github.com/Amxx/openzeppelin-contracts/tree/feature/quicksort) 
+164 −28 
[ Conversation 12](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846)
[ Commits 23](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/commits)
[ Checks 14](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/checks)
[ Files changed 7](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files)
 Open
[Arrays sorting (in memory)](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#top)
#4846
 
File filter 
 
6 / 7 files viewed
Filter changed files
 5 changes: 5 additions & 0 deletions5  
.changeset/dirty-cobras-smile.md
Viewed
  58 changes: 56 additions & 2 deletions58  
contracts/utils/Arrays.sol
Changed since last view
Viewed
@@ -12,6 +12,49 @@ import {Math} from "./math/Math.sol";
library Arrays {
    using StorageSlot for bytes32;

    /**
     * @dev Sort an array (in memory) in increasing order.
     *
     * This function does the sorting "in place", meaning that it overrides the input. The object is returned for
     * convenience, but that returned value can be discarded safely if the caller has a memory pointer to the array.
     *
     * NOTE: this function's cost is `O(n · log(n))` in average and `O(n²)` in the worst case, with n the length of the
     * array. Using it in view functions that are executed through `eth_call` is safe, but one should be very careful
     * when executing this as part of a transaction. If the array being sorted is too large, the sort operation may 
     * consume more gas than is available in a block, leading to potential DoS.
     */
    function sort(uint256[] memory array) internal pure returns (uint256[] memory) {
        _quickSort(array, 0, array.length);
        return array;
    }

    function _quickSort(uint256[] memory array, uint256 i, uint256 j) private pure {
        unchecked {
            if (j - i < 2) return;
Comment on lines +32 to +33
Member
@ernestognw ernestognw yesterday
Shouldn't this be outside of the unchecked block?

Suggested change
        unchecked {
            if (j - i < 2) return;
        if (j - i < 2) return;
        unchecked {
Otherwise one may provide i = 2 and j = 1 resulting in type(uint256).max

Member
Author
@Amxx Amxx 5 hours ago • 
This function is private. The user doesn't control the arguments i and j. The first call (done by the internal sort) uses 0 and array.length and all recursive calls after that maintain the invariant that j >= i. So no risk of overflow here, and using uncheck saves gas in a function that is recursive (so executed a lot of times)

Member
@ernestognw ernestognw Pending

Ok. We should explain this invariant in a comment if we're keeping it inside the `unchecked` blog. This is part of our guidelines:

> Unchecked arithmetic blocks should contain comments explaining why overflow is guaranteed not to happen. If the reason is immediately apparent from the line above the unchecked block, the comment may be omitted.

In my opinion, the reason is not obvious from the line above because it relies on `j` and `i` not being user-controlled, which is not implied by being `private`.
@ernestognw	Reply...

            uint256 p = i;
            for (uint256 k = i + 1; k < j; ++k) {
                if (unsafeMemoryAccess(array, i) > unsafeMemoryAccess(array, k)) {
                    _swap(array, ++p, k);
                }
            }
            _swap(array, i, p);
            _quickSort(array, i, p);
            _quickSort(array, p + 1, j);
        }
    }

    function _swap(uint256[] memory arr, uint256 i, uint256 j) private pure {
        assembly {
            let pos_i := add(add(arr, 0x20), mul(i, 0x20))
            let pos_j := add(add(arr, 0x20), mul(j, 0x20))
            let val_i := mload(pos_i)
            let val_j := mload(pos_j)
            mstore(pos_i, val_j)
            mstore(pos_j, val_i)
        }
    }

    /**
     * @dev Searches a sorted `array` and returns the first index that contains
     * a value greater or equal to `element`. If no such index exists (i.e. all
@@ -238,7 +281,7 @@ library Arrays {
     *
     * WARNING: Only use if you are certain `pos` is lower than the array length.
     */
    function unsafeMemoryAccess(uint256[] memory arr, uint256 pos) internal pure returns (uint256 res) {
    function unsafeMemoryAccess(address[] memory arr, uint256 pos) internal pure returns (address res) {
        assembly {
            res := mload(add(add(arr, 0x20), mul(pos, 0x20)))
        }
@@ -249,7 +292,18 @@ library Arrays {
     *
     * WARNING: Only use if you are certain `pos` is lower than the array length.
     */
    function unsafeMemoryAccess(address[] memory arr, uint256 pos) internal pure returns (address res) {
    function unsafeMemoryAccess(bytes32[] memory arr, uint256 pos) internal pure returns (bytes32 res) {
        assembly {
            res := mload(add(add(arr, 0x20), mul(pos, 0x20)))
        }
    }

    /**
     * @dev Access an array in an "unsafe" way. Skips solidity "index-out-of-range" check.
     *
     * WARNING: Only use if you are certain `pos` is lower than the array length.
     */
    function unsafeMemoryAccess(uint256[] memory arr, uint256 pos) internal pure returns (uint256 res) {
        assembly {
            res := mload(add(add(arr, 0x20), mul(pos, 0x20)))
        }
  4 changes: 2 additions & 2 deletions4  
scripts/generate/templates/[Checkpoints.t.js](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#diff-ad68118113721967a573f2d38eafa453504c9f3115571a539142d479accf7190)
Viewed
 15 changes: 15 additions & 0 deletions15  
test/utils/[Arrays.t.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#diff-c3a2bf0a214322cab456cc9ca3e95f7a7000d144e98005a0f3fc32cea9581b8a)
Viewed
  105 changes: 84 additions & 21 deletions105  
test/utils/[Arrays.test.js](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#diff-ba82daf17dde9aeec7fbc60c10c890c1d4a0983d01d793f6f935afcc0787e072)
Viewed
  1 change: 0 additions & 1 deletion1  
test/utils/[Base64.t.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#diff-a2dcf86ee37d3f7d7ad0fdfa13336ad65be2663aa8d75a4830367c0ee72dee03)
Viewed
  4 changes: 2 additions & 2 deletions4  
test/utils/structs/[Checkpoints.t.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4846/files#diff-b624a3b774811c2ee54d5651c187fbf436fbaa5cd94f2d9ea732b253cfefa3f3)
Viewed
Footer
© 2024 GitHub, Inc.
Footer navigation
Terms
Privacy
Security
Status
Docs
Contact
Manage cookies
Do not share my personal information

Reply:
Added comments. solving